File: trees-and-graphs/number_of_islands.py
# 200. Number of Islands
# https://leetcode.com/problems/number-of-islands/
# Difficulty: Medium
# Strategy: DFS
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def calc_neighbours(i, j):
            neighbours = []
            for dir in directions:
                i_check = i + dir[0]
                j_check = j + dir[1]

                if 0 <= i_check < rows and 0 <= j_check < cols and grid[i_check][j_check] == "1" and (i_check, j_check) not in seen:
                    seen.add((i_check, j_check))
                    neighbours.append((i_check, j_check))

            return neighbours


        def dfs(i, j):
            neighbours = calc_neighbours(i, j)
            for neighbour in neighbours:
                dfs(neighbour[0], neighbour[1])


        directions = [(-1, 0), (1, 0), (0, 1), (0, -1)]
        rows = len(grid)
        cols = len(grid[0])
        seen = set()
        ans = 0

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] != "0" and (i, j) not in seen:
                    seen.add((i, j))
                    dfs(i, j)
                    ans += 1

        return ans
    

def test():
    sol = Solution()

    grid = [
        ["1","1","1","1","0"],
        ["1","1","0","1","0"],
        ["1","1","0","0","0"],
        ["0","0","0","0","0"]
    ]

    return sol.numIslands(grid)


print(test())

File: trees-and-graphs/number_of_islands.cpp
// 200. Number of Islands
// https://leetcode.com/problems/number-of-islands/description/
// Difficulty: Medium
// Strategy: DFS
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int n;
    int m;
    vector<vector<bool>> seen;
    vector<vector<int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    vector<vector<char>> grid;

    int numIslands(vector<vector<char>>& grid) {
        this->grid = grid;
        m = grid.size();     // row
        n = grid[0].size();  // col
        seen = vector(m, vector<bool>(n, false));
        int ans = 0;

        for (int row = 0; row < m; row++) {
            for (int col = 0; col < n; col++) {
                if (grid[row][col] == '1' && !seen[row][col]) {
                    ans++;
                    seen[row][col] = true;
                    dfs(row, col);
                }
            }
        }

        return ans;
    }

    void dfs(int row, int col) {
        for (auto& dir : directions) {
            int next_row = row + dir[0];
            int next_col = col + dir[1];

            if (is_valid(next_row, next_col) && !seen[next_row][next_col]) {
                seen[next_row][next_col] = true;
                dfs(next_row, next_col);
            }
        }
    }

    bool is_valid(int row, int col) {
        return row >= 0 && row < m && col >= 0 && col < n && grid[row][col] == '1';
    }

};

int main() {
    Solution sol;

    vector<vector<char>> grid = { {'1','1','1','1','0'},
                                  {'1','1','0','1','0'},
                                  {'1','1','0','0','0'},
                                  {'0','0','0','0','0'}
                                };
    
    cout << sol.numIslands(grid) << endl;
    
    return 0;
}


